src/choose_block_size.jl

```diff
@@ -8,16 +8,23 @@ end
 
 choose_block_size(C, A, B, block_size::Integer) = block_size
 
-function choose_parameter(C, A, B, ::Nothing)
+choose_parameter(::Type{Singlethreaded}, C, A, B, block_size) =
+    Singlethreaded(choose_block_size(C, A, B, block_size))
+
+function choose_parameter(::Type{Multithreaded}, C, A, B, ::Nothing, block_size)
```

Comment:
As a minor comment here, if we aren't sure this inlines, I do prefer `::Val{T}` over `::Type{T}` to avoid passing a datatype as an argument.

Reply:
I don't *think* it's a problem when the first parameter is a constant. That said, we can also simply have `choose_multithread_parameter(C, A, B, ...)` and `choose_singlethread_parameter(C, A, B, ...)` here.

Reply:
Constant data types tend to be stack allocated and passed via a pointer.
`Val{T}()` is a singleton and thus removed as an argument altogether.
If you prefer `choose_multithreaded_parameter(...)` over `choose_parameter(Val(Multithreaded), ...)`, that's fine too.

Reply:
Hmm, I'm now not so sure when this happens.
```julia
julia> @noinline foo(::Type{T}, x) where {T} = log(T(x))
foo (generic function with 1 method)

julia> bar(x) = foo(Float32, x)
bar (generic function with 1 method)

julia> @cl bar(4)
```
```llvm
define float @julia_bar_639(i64 signext %0) #0 !dbg !5 {
top:
  %1 = call float @j_foo_641(i64 signext %0) #0, !dbg !7
  ret float %1, !dbg !7
}
```

Reply:
I did see this in some more complicated examples in the past...
But seems like some combination of a newer Julia version, and simpler examples make this a non-issue.

Reply:
Yeah, I agree passing around type can be a performance pitfall sometimes. I think I often saw this when passing a type as named argument or capturing it in a closure. I think both cases can capture types in a struct as `DataType`s. If it doesn't const-prop, it invokes a type-instability and also the `DataType`s need to be allocated somewhere.

Reply:
I split `choose_parameter` to `choose_singlethread_parameter` and `choose_multithread_parameter` to avoid dispatch on `Type` 08d944a1ba59e3a63c8f97a862baad78a4c9433b.  I don't need dispatch there anyway so I think it's cleaner.

Reply:
My example of `::Type{T}` being stack-allocated and passed as a pointer is from LoopVectorization once upon a time.
I eventually fixed this (and also tried to pass as many arguments in registers as possible), but could probably try checking out an old version to see if I can reproduce it.